modules/cuml/src/umap.ts

```diff
@@ -0,0 +1,377 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {MemoryData} from '@nvidia/cuda';
+import {DataFrame, Float32, Float64, Int64, Integral, Series} from '@rapidsai/cudf';
+import {DeviceBuffer} from '@rapidsai/rmm';
+
+import {CUMLLogLevels, MetricType} from './mappings';
+import {UMAPBase, UMAPInterface, UMAPParams} from './umap_base';
+import {dataframeToSeries, seriesToDataframe} from './utilities/array_utils';
+
+export type Numeric = Integral|Float32;
+
+export type outputType = 'dataframe'|'series'|'devicebuffer';
+
+export type returnTypeMap<T extends outputType, R extends Numeric> = {
+  'dataframe': DataFrame<{[P in number]: R}>,
+  'series': Series<R>,
+  'devicebuffer': DeviceBuffer
+}[T];
+
+export class UMAP<O extends outputType = any> {
+  public _umap: UMAPInterface;
+  public _embeddings: MemoryData|DeviceBuffer;
+  public outputType: O;
+
+  constructor(input: UMAPParams, outputType: O) {
+    this._umap       = new UMAPBase(input);
+    this._embeddings = new DeviceBuffer();
+    this.outputType  = outputType;
+  }
+
+  protected _generate_embeddings(nSamples: number, dtype: Numeric): MemoryData {
+    return Series.sequence({type: dtype, size: nSamples * this._umap.nComponents, init: 0, step: 0})
+      .data.buffer;
+  }
+
+  // throw runtime error if type if float64
+  protected _check_type(X: Numeric) {
+    if (X.compareTo(new Float64)) {
+      throw new Error('Expected input to be of type in [Integral, Float32] but got Float64');
+    }
+  }
+
+  protected _process_embeddings<T extends Numeric>(embeddings: MemoryData,
+                                                   nSamples: number,
+                                                   dtype: T): returnTypeMap<O, T> {
+    if (this.outputType == 'dataframe') {
+      return seriesToDataframe(
+               Series.new({type: dtype, data: embeddings}), nSamples, this.nComponents) as
+             returnTypeMap<O, T>;
+    } else if (this.outputType == 'series') {
+      return Series.new({type: dtype, data: embeddings}) as returnTypeMap<O, T>;
+    } else {
+      return embeddings as returnTypeMap<O, T>;
+    }
+  }
+  /**
+   * Fit X into an embedded space
+   * @param X cuDF Series containing floats or doubles in the format [x1, y1, z1, x2, y2, z2...] for
+   *   features x, y & z.
+   * @param y cuDF Series containing target values
+   * @param convertDType When set to True, the method will automatically convert the inputs to
+   *   float32
+   * @param nFeatures number of features in the input X, if X is of the format [x1,y1,x2,y2...]
+   */
+  fitSeries<T extends Series<Numeric>, R extends Series<Numeric>, B extends boolean>(
+    X: T, y: null|R, convertDType: B, nFeatures = 1) {
+    // runtime type check
+    this._check_type(X.type);
```

Comment:
so the Numeric declared here is a already custom one `(Integral|Float32)`, https://github.com/AjayThorve/rapids-js-1/blob/bindings/cuml-umap/modules/cuml/src/umap.ts#L23.

But that allowed only a complie time check. 

I thought handling the check at runtime would also be necessary as otherwise, lib-cuml throws an error which is not intuitive. Python cuml also throws a similar error if we pass in `float64`